File: tests/resolver/paths.test.ts
import * as os from "node:os";
import * as path from "node:path";

import { describe, expect, it } from "vitest";

import { InvalidManifestError } from "../../src/errors/index.js";
import {
  buildPaths,
  classifyInclude,
  isExternal,
  isValidProfileName,
} from "../../src/resolver/paths.js";

describe("classifyInclude() — R37", () => {
  const projectRoot = path.resolve("/tmp/some-project");
  const referencingDir = path.join(projectRoot, ".claude-profiles", "myprofile");
  const paths = buildPaths(projectRoot);

  it("treats bare names as components rooted at _components/<name>/", () => {
    const ref = classifyInclude("compA", referencingDir, paths, "p");
    expect(ref.kind).toBe("component");
    expect(ref.resolvedPath).toBe(
      path.join(projectRoot, ".claude-profiles", "_components", "compA"),
    );
    expect(ref.external).toBe(false);
  });

  it("treats './...' as relative to the referencing profile dir", () => {
    const ref = classifyInclude("./neighbor", referencingDir, paths, "p");
    expect(ref.kind).toBe("relative");
    expect(ref.resolvedPath).toBe(path.resolve(referencingDir, "./neighbor"));
  });

  it("treats '../...' as relative", () => {
    const ref = classifyInclude("../sib", referencingDir, paths, "p");
    expect(ref.kind).toBe("relative");
    expect(ref.resolvedPath).toBe(path.resolve(referencingDir, "../sib"));
  });

  it("expands '~/...' to the home directory and tags kind=tilde", () => {
    const ref = classifyInclude("~/some/path", referencingDir, paths, "p");
    expect(ref.kind).toBe("tilde");
    expect(ref.resolvedPath).toBe(path.resolve(path.join(os.homedir(), "some/path")));
  });

  it("expands bare '~' to the home directory", () => {
    const ref = classifyInclude("~", referencingDir, paths, "p");
    expect(ref.kind).toBe("tilde");
    expect(ref.resolvedPath).toBe(path.resolve(os.homedir()));
  });

  it("flags an absolute path outside the project root as external (kind unchanged)", () => {
    const outside = "/var/tmp/external-profile";
    const ref = classifyInclude(outside, referencingDir, paths, "p");
    expect(ref.external).toBe(true);
    expect(ref.kind).toBe("absolute"); // syntactic kind preserved
  });

  it("does NOT flag an absolute path inside the project root as external", () => {
    const inside = path.join(projectRoot, ".claude-profiles", "_components", "x");
    const ref = classifyInclude(inside, referencingDir, paths, "p");
    expect(ref.external).toBe(false);
    expect(ref.kind).toBe("absolute");
  });

  it("rejects bare-with-slashes strings (R37 admits exactly four forms)", () => {
    expect(() => classifyInclude("foo/bar", referencingDir, paths, "p")).toThrow(
      InvalidManifestError,
    );
  });

  it("rejects bare-with-backslashes strings on every platform", () => {
    expect(() => classifyInclude("foo\\bar", referencingDir, paths, "p")).toThrow(
      InvalidManifestError,
    );
  });

  it("rejects '~user' form as unsupported", () => {
    expect(() => classifyInclude("~bob/path", referencingDir, paths, "p")).toThrow(
      InvalidManifestError,
    );
  });

  it("rejects empty string", () => {
    expect(() => classifyInclude("", referencingDir, paths, "p")).toThrow(
      InvalidManifestError,
    );
  });

  it("rejects '.' and '..' (would otherwise resolve to _components/ or .claude-profiles/)", () => {
    expect(() => classifyInclude(".", referencingDir, paths, "p")).toThrow(
      InvalidManifestError,
    );
    expect(() => classifyInclude("..", referencingDir, paths, "p")).toThrow(
      InvalidManifestError,
    );
  });
});

describe("isValidProfileName()", () => {
  it("accepts a plain bare name", () => {
    expect(isValidProfileName("frontend")).toBe(true);
    expect(isValidProfileName("staging-leaf")).toBe(true);
  });

  it("rejects path traversal forms", () => {
    expect(isValidProfileName("..")).toBe(false);
    expect(isValidProfileName(".")).toBe(false);
    expect(isValidProfileName("../outside")).toBe(false);
    expect(isValidProfileName("foo/bar")).toBe(false);
    expect(isValidProfileName("foo\\bar")).toBe(false);
  });

  it("rejects empty / hidden / underscore-prefixed names", () => {
    expect(isValidProfileName("")).toBe(false);
    expect(isValidProfileName("_components")).toBe(false);
    expect(isValidProfileName(".hidden")).toBe(false);
  });
});

describe("isExternal()", () => {
  it("returns true when the path is outside the project root", () => {
    expect(isExternal("/var/data", "/Users/x/proj")).toBe(true);
  });

  it("returns false when the path is inside the project root", () => {
    expect(isExternal("/Users/x/proj/sub/dir", "/Users/x/proj")).toBe(false);
  });

  it("returns false when the path is the project root itself", () => {
    expect(isExternal("/Users/x/proj", "/Users/x/proj")).toBe(false);
  });

  it("does NOT flag a filename whose first component begins with '..' but is not a traversal segment", () => {
    // Regression guard for the prefix-vs-segment bug: `..hidden` is a
    // legitimate in-root filename, not a `..` parent-traversal segment.
    expect(isExternal("/proj/..hidden", "/proj")).toBe(false);
  });
});

File: go/internal/resolver/paths_test.go
package resolver_test

import (
	stderrors "errors"
	"os"
	"path/filepath"
	"strings"
	"testing"

	pipelineerrors "github.com/htxryan/c3p/internal/errors"
	"github.com/htxryan/c3p/internal/resolver"
)

func TestClassifyInclude_Component(t *testing.T) {
	projectRoot := mustAbs(t, "/tmp/some-project")
	referencingDir := filepath.Join(projectRoot, ".claude-profiles", "myprofile")
	paths := resolver.BuildPaths(projectRoot)

	ref, err := resolver.ClassifyInclude("compA", referencingDir, paths, "p")
	if err != nil {
		t.Fatalf("classify component: %v", err)
	}
	if ref.Kind != resolver.IncludeKindComponent {
		t.Fatalf("kind: want component, got %q", ref.Kind)
	}
	wantPath := filepath.Join(projectRoot, ".claude-profiles", "_components", "compA")
	if ref.ResolvedPath != wantPath {
		t.Fatalf("resolvedPath: want %q, got %q", wantPath, ref.ResolvedPath)
	}
	if ref.External {
		t.Fatalf("expected non-external for in-repo component")
	}
}

func TestClassifyInclude_Relative(t *testing.T) {
	projectRoot := mustAbs(t, "/tmp/some-project")
	referencingDir := filepath.Join(projectRoot, ".claude-profiles", "myprofile")
	paths := resolver.BuildPaths(projectRoot)

	ref, err := resolver.ClassifyInclude("./neighbor", referencingDir, paths, "p")
	if err != nil {
		t.Fatalf("classify ./neighbor: %v", err)
	}
	if ref.Kind != resolver.IncludeKindRelative {
		t.Fatalf("kind: want relative, got %q", ref.Kind)
	}
	want := filepath.Clean(filepath.Join(referencingDir, "neighbor"))
	if ref.ResolvedPath != want {
		t.Fatalf("resolvedPath: want %q, got %q", want, ref.ResolvedPath)
	}

	ref2, err := resolver.ClassifyInclude("../sib", referencingDir, paths, "p")
	if err != nil {
		t.Fatalf("classify ../sib: %v", err)
	}
	if ref2.Kind != resolver.IncludeKindRelative {
		t.Fatalf("kind: want relative, got %q", ref2.Kind)
	}
}

func TestClassifyInclude_TildeExpands(t *testing.T) {
	projectRoot := mustAbs(t, "/tmp/some-project")
	referencingDir := filepath.Join(projectRoot, ".claude-profiles", "myprofile")
	paths := resolver.BuildPaths(projectRoot)

	ref, err := resolver.ClassifyInclude("~/some/path", referencingDir, paths, "p")
	if err != nil {
		t.Fatalf("classify ~: %v", err)
	}
	if ref.Kind != resolver.IncludeKindTilde {
		t.Fatalf("kind: want tilde, got %q", ref.Kind)
	}
	home, _ := os.UserHomeDir()
	want := filepath.Clean(filepath.Join(home, "some/path"))
	if ref.ResolvedPath != want {
		t.Fatalf("resolvedPath: want %q, got %q", want, ref.ResolvedPath)
	}

	ref2, err := resolver.ClassifyInclude("~", referencingDir, paths, "p")
	if err != nil {
		t.Fatalf("classify bare ~: %v", err)
	}
	if ref2.Kind != resolver.IncludeKindTilde {
		t.Fatalf("kind: want tilde, got %q", ref2.Kind)
	}
}

func TestClassifyInclude_TildeInsideProjectRoot(t *testing.T) {
	// When projectRoot itself happens to live under $HOME, a tilde include
	// whose expanded path falls inside projectRoot must be flagged
	// non-external. Coverage gap surfaced by the D1 review.
	home, err := os.UserHomeDir()
	if err != nil {
		t.Skipf("no home dir: %v", err)
	}
	projectRoot := filepath.Join(home)
	referencingDir := filepath.Join(projectRoot, ".claude-profiles", "myprofile")
	paths := resolver.BuildPaths(projectRoot)

	ref, err := resolver.ClassifyInclude("~", referencingDir, paths, "p")
	if err != nil {
		t.Fatalf("classify ~: %v", err)
	}
	if ref.External {
		t.Fatalf("expected external=false when ~ resolves inside projectRoot, got external=%v", ref.External)
	}
}

func TestIsExternal_DotDotPrefixedFilename(t *testing.T) {
	// A filename whose first component begins with `..` characters but
	// is not a path-traversal segment must not be flagged external.
	// Regression guard for the prefix-vs-segment bug.
	if resolver.IsExternal("/proj/..hidden", "/proj") {
		t.Fatalf(`"/proj/..hidden" must not be external`)
	}
}

func TestClassifyInclude_AbsoluteExternal(t *testing.T) {
	projectRoot := mustAbs(t, "/tmp/some-project")
	referencingDir := filepath.Join(projectRoot, ".claude-profiles", "myprofile")
	paths := resolver.BuildPaths(projectRoot)

	outside := "/var/tmp/external-profile"
	ref, err := resolver.ClassifyInclude(outside, referencingDir, paths, "p")
	if err != nil {
		t.Fatalf("classify abs external: %v", err)
	}
	if !ref.External {
		t.Fatalf("expected external=true for /var/tmp/...")
	}
	if ref.Kind != resolver.IncludeKindAbsolute {
		t.Fatalf("kind: want absolute, got %q", ref.Kind)
	}
}

func TestClassifyInclude_AbsoluteInternal(t *testing.T) {
	projectRoot := mustAbs(t, "/tmp/some-project")
	referencingDir := filepath.Join(projectRoot, ".claude-profiles", "myprofile")
	paths := resolver.BuildPaths(projectRoot)

	inside := filepath.Join(projectRoot, ".claude-profiles", "_components", "x")
	ref, err := resolver.ClassifyInclude(inside, referencingDir, paths, "p")
	if err != nil {
		t.Fatalf("classify abs internal: %v", err)
	}
	if ref.External {
		t.Fatalf("expected external=false for in-repo absolute")
	}
	if ref.Kind != resolver.IncludeKindAbsolute {
		t.Fatalf("kind: want absolute, got %q", ref.Kind)
	}
}

func TestClassifyInclude_RejectsBareWithSlashes(t *testing.T) {
	projectRoot := mustAbs(t, "/tmp/some-project")
	referencingDir := filepath.Join(projectRoot, ".claude-profiles", "myprofile")
	paths := resolver.BuildPaths(projectRoot)

	_, err := resolver.ClassifyInclude("foo/bar", referencingDir, paths, "p")
	var ime *pipelineerrors.InvalidManifestError
	if !stderrors.As(err, &ime) {
		t.Fatalf("expected InvalidManifestError, got %v", err)
	}
}

func TestClassifyInclude_RejectsBareWithBackslashes(t *testing.T) {
	projectRoot := mustAbs(t, "/tmp/some-project")
	referencingDir := filepath.Join(projectRoot, ".claude-profiles", "myprofile")
	paths := resolver.BuildPaths(projectRoot)

	_, err := resolver.ClassifyInclude(`foo\bar`, referencingDir, paths, "p")
	var ime *pipelineerrors.InvalidManifestError
	if !stderrors.As(err, &ime) {
		t.Fatalf("expected InvalidManifestError, got %v", err)
	}
}

func TestClassifyInclude_RejectsTildeUserForm(t *testing.T) {
	projectRoot := mustAbs(t, "/tmp/some-project")
	referencingDir := filepath.Join(projectRoot, ".claude-profiles", "myprofile")
	paths := resolver.BuildPaths(projectRoot)

	_, err := resolver.ClassifyInclude("~bob/path", referencingDir, paths, "p")
	var ime *pipelineerrors.InvalidManifestError
	if !stderrors.As(err, &ime) {
		t.Fatalf("expected InvalidManifestError, got %v", err)
	}
	if !strings.Contains(err.Error(), "~user") {
		t.Fatalf("error message must mention ~user: %q", err.Error())
	}
}

func TestClassifyInclude_RejectsEmpty(t *testing.T) {
	projectRoot := mustAbs(t, "/tmp/some-project")
	referencingDir := filepath.Join(projectRoot, ".claude-profiles", "myprofile")
	paths := resolver.BuildPaths(projectRoot)

	_, err := resolver.ClassifyInclude("", referencingDir, paths, "p")
	var ime *pipelineerrors.InvalidManifestError
	if !stderrors.As(err, &ime) {
		t.Fatalf("expected InvalidManifestError, got %v", err)
	}
}

func TestIsValidProfileName(t *testing.T) {
	cases := []struct {
		name string
		want bool
	}{
		{"frontend", true},
		{"staging-leaf", true},
		{"foo123", true},
		{"..", false},
		{".", false},
		{"../outside", false},
		{"foo/bar", false},
		{`foo\bar`, false},
		{"", false},
		{"_components", false},
		{".hidden", false},
		{"CON", false},
		{"con.txt", false},
		{"COM1", false},
		{"LPT9", false},
		{"trailing.", false},
		{"trailing ", false},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := resolver.IsValidProfileName(tc.name); got != tc.want {
				t.Fatalf("IsValidProfileName(%q): want %v, got %v", tc.name, tc.want, got)
			}
		})
	}
}

func TestIsExternal(t *testing.T) {
	cases := []struct {
		absPath, root string
		want          bool
	}{
		{"/var/data", "/Users/x/proj", true},
		{"/Users/x/proj/sub/dir", "/Users/x/proj", false},
		{"/Users/x/proj", "/Users/x/proj", false},
	}
	for _, tc := range cases {
		t.Run(tc.absPath, func(t *testing.T) {
			if got := resolver.IsExternal(tc.absPath, tc.root); got != tc.want {
				t.Fatalf("IsExternal(%q, %q): want %v, got %v", tc.absPath, tc.root, tc.want, got)
			}
		})
	}
}

func mustAbs(t *testing.T, p string) string {
	t.Helper()
	abs, err := filepath.Abs(p)
	if err != nil {
		t.Fatalf("abs: %v", err)
	}
	return abs
}
